sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala

```diff
@@ -963,98 +964,178 @@ class StatisticsSuite extends StatisticsCollectionTestBase with TestHiveSingleto
-        "spark.sql.statistics.colStats.cint.max" -> "4",
-        "spark.sql.statistics.colStats.cint.maxLen" -> "4",
-        "spark.sql.statistics.colStats.cint.min" -> "1",
-        "spark.sql.statistics.colStats.cint.nullCount" -> "1",
-        "spark.sql.statistics.colStats.cint.version" -> "1",
-        "spark.sql.statistics.colStats.clong.avgLen" -> "8",
-        "spark.sql.statistics.colStats.clong.distinctCount" -> "2",
-        "spark.sql.statistics.colStats.clong.max" -> "5",
-        "spark.sql.statistics.colStats.clong.maxLen" -> "8",
-        "spark.sql.statistics.colStats.clong.min" -> "1",
-        "spark.sql.statistics.colStats.clong.nullCount" -> "1",
-        "spark.sql.statistics.colStats.clong.version" -> "1",
-        "spark.sql.statistics.colStats.cshort.avgLen" -> "2",
-        "spark.sql.statistics.colStats.cshort.distinctCount" -> "2",
-        "spark.sql.statistics.colStats.cshort.max" -> "3",
-        "spark.sql.statistics.colStats.cshort.maxLen" -> "2",
-        "spark.sql.statistics.colStats.cshort.min" -> "1",
-        "spark.sql.statistics.colStats.cshort.nullCount" -> "1",
-        "spark.sql.statistics.colStats.cshort.version" -> "1",
-        "spark.sql.statistics.colStats.cstring.avgLen" -> "3",
-        "spark.sql.statistics.colStats.cstring.distinctCount" -> "2",
-        "spark.sql.statistics.colStats.cstring.maxLen" -> "3",
-        "spark.sql.statistics.colStats.cstring.nullCount" -> "1",
-        "spark.sql.statistics.colStats.cstring.version" -> "1",
-        "spark.sql.statistics.colStats.ctimestamp.avgLen" -> "8",
-        "spark.sql.statistics.colStats.ctimestamp.distinctCount" -> "2",
-        "spark.sql.statistics.colStats.ctimestamp.max" -> "2016-05-09 00:00:02.0",
-        "spark.sql.statistics.colStats.ctimestamp.maxLen" -> "8",
-        "spark.sql.statistics.colStats.ctimestamp.min" -> "2016-05-08 00:00:01.0",
-        "spark.sql.statistics.colStats.ctimestamp.nullCount" -> "1",
-        "spark.sql.statistics.colStats.ctimestamp.version" -> "1"
-      ))
+        checkColStatsProps(expectedSerializedColStats ++ expectedSerializedHistograms)
+      }
+    }
+  }
+
+  test("serde/deser of histograms exceeding 4000 length") {
+    import testImplicits._
+
+    def checkBucketsOrder(buckets: Array[HistogramBucket]): Unit = {
+      for (i <- buckets.indices) {
+        val b = buckets(i)
+        assert(b.lo <= b.hi)
+        if (i > 0) {
+          val pre = buckets(i - 1)
+          assert(pre.hi <= b.lo)
+        }
+      }
+    }
+
+    val startTimestamp = DateTimeUtils.fromJavaTimestamp(Timestamp.valueOf("2016-05-08 00:00:01"))
+    val df = (1 to 5000)
+      .map(i => (i, DateTimeUtils.toJavaTimestamp(startTimestamp + i)))
+      .toDF("cint", "ctimestamp")
+    val tableName = "long_histogram_test"
+
+    withTable(tableName) {
+      df.write.saveAsTable(tableName)
+
+      withSQLConf(
+        SQLConf.HISTOGRAM_ENABLED.key -> "true", SQLConf.HISTOGRAM_BUCKETS_NUM.key -> "1000") {
```

Comment:
If users set 1000 buckets, I think it's more reasonable to just fail... My point is, it doesn't worth this complexity to this corner case.